### ISSUE TITLE
[@sls-next/lambda-at-edge] Build-time environment variables not propagated to dependencies w/ useServerlessTraceTarget

### ISSUE THREAD

**Issue by ramosbugs** [serverless-nextjs/serverless-next.js#790]
### Describe the bug

By default (i.e., without `useServerlessTraceTarget: true`), Next.JS [uses the `ServerlessPlugin`](https://github.com/vercel/next.js/blob/8ff52136146b3276bf53d40c7ee1cfe7f0c9647f/packages/next/build/webpack-config.ts#L1069) to [inline all dependencies](https://github.com/vercel/next.js/blob/8ff52136146b3276bf53d40c7ee1cfe7f0c9647f/packages/next/build/webpack/plugins/serverless-plugin.ts#L38) into each page's entry point bundle. The environment variables [set using the `DefinePlugin`](https://github.com/vercel/next.js/blob/8ff52136146b3276bf53d40c7ee1cfe7f0c9647f/packages/next/build/webpack-config.ts#L929-L1031) are then inlined in both the page source and all of its dependencies.

When using `useServerlessTraceTarget: true`, the `ServerlessPlugin` is skipped, and Webpack doesn't inline the dependencies. Instead, the `lambda-at-edge` builder [copies the relevant `node_modules`](https://github.com/serverless-nextjs/serverless-next.js/blob/f7300978926c1225a0024063bd67de56961e970a/packages/libs/lambda-at-edge/src/build.ts#L268-L273). These files are left untouched by Webpack, so they continue to have references like `process.env.NODE_ENV` in the code.

Unfortunately, nothing sets these environment variables at runtime, including the often-used `NODE_ENV`. This leads to a pretty big semantic difference between the `experimental-serverless-trace` build and the old `serverless` build. The old one had all of its environment variables folded in at build time, while the new one expects the environment variables to be present at runtime (which they aren't by default).

### Actual behavior

Environment variables are not propagated to dependencies when using `useServerlessTraceTarget: true`.

### Expected behavior

Environment variables should have the same behavior with and without `useServerlessTraceTarget: true`.

### Steps to reproduce

1. Build a Next.JS project that has an NPM dependency that relies on an environment variable like `process.env.NODE_ENV` with `useServerlessTraceTarget: true`.
2. Deploy the project to Lambda@Edge.
3. Visit the page and observe from the logs or the page's behavior that `process.env.NODE_ENV` is `undefined` instead of `production`.
4. Rebuild without `useServerlessTraceTarget: true` and re-deploy.
5. Visit the page and observe that `process.env.NODE_ENV` has the value `production` (since it was inlined at build time).

### Versions

- OS/Environment: Darwin
- @sls-next/lambda-at-edge version: 1.8.0-alpha.4
- Next.js version: 10.0.1

### Additional context

It seems like there are two rough approaches to solving the problem:
1. Build time: run some sort of Webpack build over the traced dependencies to inline any environment variables within `node_modules`. This would make `useServerlessTraceTarget: true` behave the same as the old approach.
2. Runtime: save the environment variables from build time and set them (early on) at runtime so that they're present if any dependencies need them. This behavior would still differ from the old approach, since `process.env` would contain more values at runtime than it did with the old approach.

### Checklist

- [X] You have reviewed the README and [FAQs](https://github.com/serverless-nextjs/serverless-next.js#faq), which answers several common questions.
- [X] Please first try using the latest `@sls-next/serverless-component` [release version](https://github.com/serverless-nextjs/serverless-next.js/releases), which may have already fixed your issue. Note that the old `serverless-next.js` component and the `serverless-next.js` plugin are deprecated and no longer maintained.

**Comment by dphang** [comment]
Yes, I had this problem with serverless-trace with material-ui (though I don't use it anymore, due to unrelated perf reasons), found it because it was trying to use development mode (causing style class mismatches). Though if I used an `.env` file and set `NODE_ENV` to `production`, I believe it properly inlined the `NODE_ENV` into material-ui.

**Comment by ramosbugs** [comment]
sounds like we ran into the exact same issue then!

> Though I used an `.env` file and explicitly set `NODE_ENV` to `production`, I believe it properly inlined the `NODE_ENV` into material-ui.

yup, adding a `.env` to set `NODE_ENV` fixes that specific issue. thanks for the suggestion! it looks like `.env` environment vars get [loaded at runtime](https://github.com/vercel/next.js/blob/415d7fd75ebfca4dc796bb7892a348bab3fa4920/packages/next/build/webpack/loaders/next-serverless-loader.ts#L145) rather than inlined during the build.

still though, this behavior feels like a bit of a rough edge. if we think about the `DefinePlugin` inlining as just an optimization in which the environment variable is still expected to be present at runtime, then the difference in behavior between `serverless` and `experimental-serverless-trace` seems totally fine. in both cases, it seems like an error for any of those environment variables to be missing or to have different values at runtime since in either case the built-time optimization would have altered the runtime behavior.

from the link above, it appears that Next.JS already loads the `.env` values again at runtime. however, the `config.env` (from `next.config.js`) and build-time environment variables aren't similarly copied into the runtime environment. it feels like all of those sources of `DefinePlugin` environment variables should be loaded automatically again at runtime. I guess that's more of a Next.JS issue than a serverless-next.js issue, though.

thoughts?

**Comment by ratkorle** [comment]
I have similar issue.
If you can please have a look:
[https://github.com/serverless-nextjs/serverless-next.js/issues/887](https://github.com/serverless-nextjs/serverless-next.js/issues/887)

**Comment by jonahallibone** [comment]
I am using Codebuild to output my project, using `useServerlessTraceTarget`, and I've noticed the env variables set in the Codebuild env configuration work fine for the frontend, but do not get inlined in the api pages. Is this related to this issue? I believe it is but I haven't seen any mention of CI tools being used.
